st2client/tests/unit/test_auth.py

```diff
@@ -41,277 +41,393 @@
+
+    def __init__(self, *args, **kwargs):
+        super(TestLoginWritePwdOkay, self).__init__(
+            config_file=self.CONFIG_FILE, *args, **kwargs
+        )
 
     @mock.patch.object(
         requests, 'post',
-        mock.MagicMock(return_value=base.FakeResponse(json.dumps({}), 200, 'OK')))
-    @mock.patch("st2client.commands.auth.ConfigParser")
-    @mock.patch("st2client.commands.auth.open")
-    @mock.patch("st2client.commands.auth.BaseCLIApp")
-    @mock.patch("st2client.commands.auth.getpass")
-    def test_whoami_missing_credentials_section(self, mock_gp, mock_cli, mock_open, mock_cfg):
-        """Test 'st2 whoami' functionality with a missing credentials section
-        """
-
-        # mocked config that is empty (no credentials section at all)
-        mock_cfg.return_value.read.return_value = {}
-
-        # Mock config
-        config_file = config_parser.ST2_CONFIG_PATH
-        self.shell._get_config_file_path = mock.MagicMock(return_value="/tmp/st2config")
-        mock_cli.return_value._get_config_file_path.return_value = config_file
-
-        # Trigger keyerror exception when trying to access username.
-        # We have to do it this way because ConfigParser acts like a
-        # dict but also has methods like read()
-        attrs = {'__getitem__.side_effect': KeyError}
-        mock_cfg.return_value.configure_mock(**attrs)
-
-        # assert that the config field lookup caused the CLI to return an error code
-        # we are also using "--debug" flag to ensure the exception is re-raised once caught
-        self.assertEqual(
-            self.shell.run(['--debug', 'whoami']),
-            1
-        )
+        mock.MagicMock(return_value=base.FakeResponse(json.dumps(TOKEN), 200, 'OK')))
+    @mock.patch('st2client.commands.auth.getpass')
+    def runTest(self, mock_gp):
+        '''Test 'st2 login' functionality with --write-password flag set
+        '''
+
+        expected_username = self.TOKEN['user']
+        args = ['--config', self.CONFIG_FILE, 'login', expected_username, '--password',
+                'Password1!', '--write-password']
+
+        self.shell.run(args)
+
+        with open(self.CONFIG_FILE, 'r') as config_file:
+
+            for line in config_file.readlines():
+
+                # Make sure certain values are not present
+                self.assertFalse('olduser' in line)
 
-        # An additional assert to ensure things are being called correctly
-        mock_cfg.return_value.__getitem__.assert_called_with('credentials')
+                # Make sure configured username is what we expect
+                if 'username' in line:
+                    self.assertEquals(line.split(' ')[2][:-1], expected_username)
 
+            # validate token was created
+            self.assertTrue(os.path.isfile('%stoken-%s' % (self.DOTST2_PATH, expected_username)))
 
-class TestLogin(base.BaseCLITestCase):
+
+class TestLoginUncaughtException(TestLoginBase):
+
+    CONFIG_FILE = '/tmp/logintest.cfg'
```

Comment:
Might be a bit safer to use a random temporary path (e.g. mkstemp), besides that, LGTM.